### PR TITLE
update post process with num_beams and do_sample

### PR DESCRIPTION
**Pull request by zhenwei-intel**
## Type of Change

feature
- greedy search
- topk and topp sampling
- beam search api ready, waiting @zhentaoyu to implement

## Description

support num_beams and do_sample in generate api:

- greedy search
```python
outputs = model.generate(inputs, streamer=streamer, max_new_tokens=100, num_beams=1, do_sample=False)
```
```bash
AVX:1 AVX2:1 AVX512F:1 AVX_VNNI:0 AVX512_VNNI:1 AMX_INT8:0 AMX_BF16:0 AVX512_BF16:0 AVX512_FP16:0
beam_size: 1, do_sample: 0, top_k: 40, top_p: 0.950000
model.cpp: loading model from /mnt/disk2/data/zhenweil/codes/ggml/mpt_ne.bin
init: n_vocab    = 50432
init: n_ctx      = 512
init: n_embd     = 4096
init: n_mult     = 4096
init: n_head     = 32
init: n_layer    = 32
init: n_rot      = 32
init: n_ff       = 16384
init: n_parts    = 1
load: ne ctx size = 25365.06 MB
load: mem required  = 33557.06 MB (+ memory per state)
..................................................................................................
model_init_from_file: kv self size  =  128.00 MB
Once upon a time, there existed a little girl who loved to play with her dolls. She would dress them up in pretty clothes and put them to bed. She would feed them and play with them. She would even take them to the park and play with them.
One day, the little girl decided to take her dolls to the park. She put them in the stroller and pushed them around. She had a great time playing with them.
After a while, the little girl got tired and decided to go home. She
```
- topk and topp sampling
```python
outputs = model.generate(inputs, streamer=streamer, max_new_tokens=100, num_beams=1, do_sample=True, top_k=40, top_p=0.95)
```
```bash
AVX:1 AVX2:1 AVX512F:1 AVX_VNNI:0 AVX512_VNNI:1 AMX_INT8:0 AMX_BF16:0 AVX512_BF16:0 AVX512_FP16:0
beam_size: 1, do_sample: 1, top_k: 40, top_p: 0.950000
model.cpp: loading model from /mnt/disk2/data/zhenweil/codes/ggml/mpt_ne.bin
init: n_vocab    = 50432
init: n_ctx      = 512
init: n_embd     = 4096
init: n_mult     = 4096
init: n_head     = 32
init: n_layer    = 32
init: n_rot      = 32
init: n_ff       = 16384
init: n_parts    = 1
load: ne ctx size = 25365.06 MB
load: mem required  = 33557.06 MB (+ memory per state)
..................................................................................................
model_init_from_file: kv self size  =  128.00 MB
Once upon a time, there existed a little girl named Alice. She lived in a house with her mother, father and sister, who was called Lucy.
One day, Alice’s parents decided to go on holiday for a week. They left Alice and Lucy at home with their nanny, Mrs. Brown.
Alice and Lucy were very excited about the holiday. They had never been away from home before.
The first day of the holiday was very exciting. Alice and Lucy went to the park with Mrs. Brown. They played 
```

### CODE REVIEW

File: intel_extension_for_transformers/llm/runtime/graph/application/common.cpp [hunk 1]
```diff
@@ -561,11 +561,12 @@ gpt_vocab::id gpt_sample_top_k_top_p(const gpt_vocab& vocab, const float* logits
...
 
-gpt_vocab::id gpt_sample_top_k_top_p_repeat(const gpt_vocab& vocab, const float* logits,
+gpt_vocab::id gpt_sample_top_k_top_p_repeat(const float* logits,
                                             const int32_t* last_n_tokens_data, size_t last_n_tokens_data_size,
                                             int top_k, double top_p, double temp, int repeat_last_n,
-                                            float repeat_penalty, std::mt19937& rng) {
-  int n_logits = vocab.id_to_token.size();
+                                            float repeat_penalty) {
+  int n_logits = 50432;
+  std::mt19937 rng(1234);
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/runtime/graph/application/common.cpp:569]: for debugging?

**Comment by airMeng** [comment]
add acc and per stats in description

File: intel_extension_for_transformers/llm/runtime/graph/application/common.cpp (hunk 1 above)
Comment (zhenwei-intel) [reply intel_extension_for_transformers/llm/runtime/graph/application/common.cpp:569]: still draft

File: intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp [hunk 2]
```diff
@@ -195,39 +200,131 @@ std::vector<int> Model::generate_tokens(const std::vector<int>& input_ids) {
...
+
+    cumsum = 1.0 / cumsum;
+    for (int i = 0; i < (int)probs.size(); i++) {
+      probs[i] *= cumsum;
+    }
+  }
+
+  //    printf("\n");
+  //    for (int i = 0; i < (int) probs.size(); i++) {
+  //    for (int i = 0; i < 10; i++) {
```
Comment (a32543254) [inline intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp:308]: remove unused code

File: intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_config.h [hunk 3]
```diff
@@ -41,6 +41,7 @@ struct gpt_params {
   int32_t n_gpu_layers = 0;  // number of layers to store in VRAM
 
   // sampling parameters
+  bool do_sample = false;
```
Comment (a32543254) [inline intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_config.h:44]: default is false ?
what post process we will use as default ?

**Review by a32543254 — APPROVED** [review]
LGTM

File: intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_config.h (hunk 3 above)
Comment (zhenwei-intel) [reply intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_config.h:44]: greedy search by default

File: intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp [hunk 4]
```diff
@@ -195,39 +200,131 @@ std::vector<int> Model::generate_tokens(const std::vector<int>& input_ids) {
...
+  //    }
+
+  std::discrete_distribution<> dist(probs.begin(), probs.end());
+  int idx = dist(rng);
+
+  return logits_id[idx].second;
+}
+
+int Model::post_process(float* logits) {
+  if (params.beam_size > 1) {
```
Comment (zhentaoyu) [inline intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp:319]: `if (params.beam_size > 1 && !do_sample) `

File: intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp [hunk 5]
```diff
@@ -195,39 +200,131 @@ std::vector<int> Model::generate_tokens(const std::vector<int>& input_ids) {
...
-  int alpha_frequency = 0;
-  int alpha_presence = 0;
+int Model::post_greedy_search(float* logits) {
+  int id = std::max_element(logits, logits + n_vocab) - logits;
+  return id;
+}
+
+int Model::post_beam_search(float* logits) {
+  // TODO: to implement
+  fprintf(stderr, "beam search is not supported!\n");
```
Comment (zhentaoyu) [inline intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp:210]: due to `kv cache update func ptr` and `batch inference` implementations in different models. We may need to add a supported table here.

**Review by zhentaoyu — APPROVED** [review]

File: intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp (hunk 2 above)
Comment (zhenwei-intel) [reply intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp]: removed

File: intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp (hunk 4 above)
Comment (zhentaoyu) [reply intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp:314]: generation mode dispatch reference: https://github.com/huggingface/transformers/blob/main/src/transformers/generation/utils.py#L849-L887

Comment (zhenwei-intel) [reply intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp:314]: could this work?
```cpp
int Model::post_process(float* logits) {
  if (params.do_sample) {
    return post_sample_top_k_top_p_repeat(logits);
  }

  if (params.beam_size > 1) {
    return post_beam_search(logits);
  }

  return post_greedy_search(logits);
}
```

Comment (zhentaoyu) [reply intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp:314]: ok

Comment (zhenwei-intel) [reply intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp]: thanks, updated

File: intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp (hunk 5 above)
Comment (zhenwei-intel) [reply intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp]: update it when you do the beam search

**Pull request merged**